src/libpriv/rpmostree-core.c

```diff
@@ -2696,38 +2828,54 @@ rpmostree_context_assemble_tmprootfs (RpmOstreeContext      *self,
         return FALSE;
     }
 
+  /* We're completely disregarding how rpm normally does upgrades/downgrade here. rpm
+   * usually calculates the fate of each file and then installs the new files first, then
+   * removes the obsoleted files. So the TR_ADDED element for the new pkg comes *before* the
+   * TR_REMOVED element for the old pkg. This makes sense when operating on the live system.
+   * Though that's not the case here, so let's just take the easy way out and first nuke
+   * TR_REMOVED pkgs, then checkout the TR_ADDED ones. This will probably start to matter
+   * though when we start handling e.g. file triggers. */
+
   for (guint i = 0; i < n_rpmts_elements; i++)
     {
       rpmte te = rpmtsElement (ordering_ts, i);
       rpmElementType type = rpmteType (te);
 
       if (type == TR_ADDED)
-        {
-          DnfPackage *pkg = (void*)rpmteKey (te);
-          if (pkg == filesystem_package)
-            continue;
+        continue;
+      g_assert (type == TR_REMOVED);
```

Comment:
I'd also `g_assert_cmpint()` - this one is very unlikely to fire, but from past experience having the values in the string is *really* useful if the assertion triggers.